### PR TITLE
MONGOID-4906 Implement $type

### PR DESCRIPTION
**Pull request by ksadoff**

### CODE REVIEW

File: lib/mongoid/matcher/type.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    # @see https://docs.mongodb.com/manual/reference/operator/query/type/
+    #
+    # @api private
+    module Type
+      module_function def matches?(exists, value, condition)
+        case condition
+        when Array
+          condition.any? do |v|
+            condition = v
```
Comment (p-mongo) [inline lib/mongoid/matcher/type.rb:12]: This doesn't look right.

File: lib/mongoid/matcher/type.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+          end
+        when 1
+          # Double
+          Float === value
+        when 2
+          # String
+          String === value
+        when 3
+          # Object
+          Object === value
```
Comment (p-mongo) [inline lib/mongoid/matcher/type.rb:22]: This should probably be `Hash` and not `Object`.

File: lib/mongoid/matcher/type.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+          BSON::ObjectId === value
+        when 8
+          # Boolean
+          TrueClass === value || FalseClass === value
+        when 9
+          # Date
+          Date === value || Time === value || DateTime === value
+        when 10
+          # Null
+          BSON::NilClass === value
```
Comment (p-mongo) [inline lib/mongoid/matcher/type.rb:43]: Just `NilClass`

**Review by p-mongo — APPROVED** [review]
I added a few test cases and fixed the following:

- arrays with multiple elements as $type arguments
- null type when the field being checked is not present
- accept Ruby integers in int32/int64 matches if they are in the right range

File: spec/integration/matcher_operator_data/type_object.yml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+- name: existing field - matches ObjectId
```
Comment (ksadoff) [inline spec/integration/matcher_operator_data/type_object.yml:1]: I think the filename for this and `type_object_id.yml` were swapped accidentally.

**Review by p-mongo — APPROVED** [review]

**Pull request merged**